### PR TITLE
Fix crashes during CASE establishment.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
There were two potential sources of crashes:

1) TryNextResult on the resolver scheduled an async task that could not be canceled.  If, while that task was pending, the OperationalSessionSetup was destroyed (which could happen if another session for the same peer had become active in the meantime and another connection attempt happened) we would end up with use-after-free.

2) ScheduleSessionSetupReattempt scheduled a timer that was never being canceled.  Similar to above, if the OperationalSessionSetup got destroyed before the timer fired we could end up with use-after-free.

The fix for the first problem is to make TryNextResult synchronous, instead of scheduling an async task.

The fix for the second problem is to introduce a new state of OperationalSessionSetup that's used while waiting for the timer and to ensure that the timer is canceled when leaving that state or if the OperationalSessionSetup is destroyed.

To preserve existing behavior, if we are in the "waiting for timer" state and new connection attempt happens and there is a session we can attach to, then we immediately move to the Connected state (and cancel the timer).

### CODE REVIEW

File: src/app/OperationalSessionSetup.cpp [hunk 1]
```diff
@@ -553,27 +576,38 @@ CHIP_ERROR OperationalSessionSetup::ScheduleSessionSetupReattempt(System::Clock:
...
-
-    CHIP_ERROR err = CHIP_NO_ERROR;
+    // If we can't get a system layer, there is no way for us to cancel things
+    // at this point, but hopefully that's because everything is torn down
+    // anyway and hence the timer will not fire.
+    auto * sessionManager = mInitParams.exchangeMgr->GetSessionManager();
+    if (!sessionManager)
+    {
+        return;
+    }
```
Comment (vivien-apple) [inline src/app/OperationalSessionSetup.cpp:588]: nit: This is pure code style preference, but I do prefer the `VerifyOrReturn(sessionManager != nullptr);` style.

File: src/app/OperationalSessionSetup.cpp [hunk 2]
```diff
@@ -553,27 +576,38 @@ CHIP_ERROR OperationalSessionSetup::ScheduleSessionSetupReattempt(System::Clock:
...
 
-    if (self->mState != State::NeedsAddress)
+    auto * systemLayer = sessionManager->SystemLayer();
+    if (!systemLayer)
     {
-        err = CHIP_ERROR_INCORRECT_STATE;
+        return;
```
Comment (vivien-apple) [inline src/app/OperationalSessionSetup.cpp:593]: nit: This is pure code style preference, but I do prefer the `VerifyOrReturn(systemLayer != nullptr);` style.

File: src/app/OperationalSessionSetup.cpp [hunk 3]
```diff
@@ -553,27 +576,38 @@ CHIP_ERROR OperationalSessionSetup::ScheduleSessionSetupReattempt(System::Clock:
...
+    CHIP_ERROR err = self->LookupPeerAddress();
+    if (err == CHIP_NO_ERROR)
     {
-        self->MoveToState(State::ResolvingAddress);
-        err = self->LookupPeerAddress();
-        if (err == CHIP_NO_ERROR)
-        {
-            return;
-        }
+        return;
```
Comment (vivien-apple) [inline src/app/OperationalSessionSetup.cpp:607]: nit: `ReturnOnFailure(self->LookupPeerAddress());` ?

**Review by vivien-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

File: src/app/OperationalSessionSetup.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/OperationalSessionSetup.cpp:588]: Changed.

File: src/app/OperationalSessionSetup.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/app/OperationalSessionSetup.cpp:593]: Changed.

File: src/app/OperationalSessionSetup.cpp (hunk 3 above)
Comment (bzbarsky-apple) [reply src/app/OperationalSessionSetup.cpp:607]: No, we want to return if `LookupPeerAddress` _succeeds_.  On failure we want to execute other logic.  Leaving this as-is.

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**